compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1491,8 +1570,12 @@ class TreeUnpickler(reader: TastyReader,
                   NoDenotation
 
               val denot =
-                val d = ownerTpe.decl(name).atSignature(sig, target)
-                (if !d.exists then lookupInSuper else d).asSeenFrom(prefix)
+                if owner.is(JavaAnnotation) && name == nme.CONSTRUCTOR then
+                  // #19951 Fix up to read TASTy produced before 3.5.0 -- ignore the signature
+                  ownerTpe.decl(name).asSeenFrom(prefix)
```

Comment:
Java classes get a dummy extra constructor when parsing from sources:
https://github.com/scala/scala3/blob/04ada792f2cc81177c2f430e1a1b9060eab7b54c/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala#L124-L131
So I wonder if this check is enough in the case where we're unpickling from tasty a reference to a java annotation that is also being passed to the compiler as source in the current run (this would be good to test anyway). The dummy constructor has the flag `PrivateLocal` set so we could filter that out explicitly.

Reply:
Hum. I'm happy to filter out `PrivateLocal`. That said I'm not sure how we can construct such a scenario in practice. Since the Scala class uses the annotation, it must be compiled together with or later than the annotation. But then how would we have the Scala class from TASTy but the Java class from source? Perhaps in some complicated incremental compilation scenario with mixed compilation? :thinking: